msbuild/custom-task-code-generation/README.md

```diff
@@ -0,0 +1,370 @@
+        <SettingClass Condition="'$(SettingClass)' == ''">MySetting</SettingClass>
+        <SettingNamespace Condition="'$(SettingNamespace)' == ''">example</SettingNamespace>
+        <SettingExtensionFile Condition="'$(SettingExtensionFile)' == ''">mysettings</SettingExtensionFile>
+    </PropertyGroup>
+</Project>
+```
+
+Beyond the [build properties](https://docs.microsoft.com/visualstudio/msbuild/walkthrough-using-msbuild#build-properties) defined, actually, an important part of this file is the task registration, MSBuild must know how to locate and run the assembly that contains the task class. Tasks are registered using the [UsingTask element (MSBuild)](https://docs.microsoft.com/visualstudio/msbuild/usingtask-element-msbuild). TaskName is the name of the task to reference from the assembly. This attribute should always specify full namespaces. AssemblyFile is the file path of the assembly.
+
+The _AppSettingStronglyTyped.props_ will be automatically included when the package is installed, then our client has the task available and some default values. However, it is never used. In order to put this code in action we need to define some targets on _AppSettingStronglyTyped.targets_ file which also will be also automatically included when the package is install:
+
+```xml
+<?xml version="1.0" encoding="utf-8" ?>
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+    <!--Defining all the text files input parameters-->
+    <ItemGroup>
+        <SettingFiles Include="$(RootFolder)\*.$(SettingExtensionFile)" />
+    </ItemGroup>
+
+    <!--It is generated a target which is executed before the compilation-->
+    <Target Name="BeforeCompile" Inputs="@(SettingFiles)" Outputs="$(RootFolder)\$(SettingClass).generated.cs">
+        <!--Calling our custom task-->
+        <AppSettingStronglyTyped SettingClassName="$(SettingClass)" SettingNamespaceName="$(SettingNamespace)" SettingFiles="@(SettingFiles)">
+            <Output TaskParameter="ClassNameFile" PropertyName="SettingClassFileName" />
+        </AppSettingStronglyTyped>
+        <!--Our generated file is included to be compiled-->
+        <ItemGroup>
+            <Compile Remove="$(SettingClassFileName)" />
+            <Compile Include="$(SettingClassFileName)" />
+        </ItemGroup>
+    </Target>
+
+    <!--The generated file is deleted after a general clean. It will force the regeneration on rebuild-->
+    <Target Name="AfterClean">
+        <Delete Files="$(RootFolder)\$(SettingClass).generated.cs" />
+    </Target>
+</Project>
+```
+
+The first step is the creation of an [InputGroup](https://docs.microsoft.com/visualstudio/msbuild/msbuild-items) which represents the text files (there could be more than one) to read and it will be some of our task parameters. There are default for the location and the extension where we look for, but you can override the values defining the properties on the client MSBuild project file.
```

Comment:
```suggestion
The first step is the creation of an [InputGroup](https://docs.microsoft.com/visualstudio/msbuild/msbuild-items) which represents the text files (there could be more than one) to read and it will be some of our task parameters. There are defaults for the location and the extension to search for, but you can override the values defining the properties on the client MSBuild project file.
```